### DISCUSSION TITLE
Oqtane Security

### DISCUSSION THREAD

**Discussion by rossmcphee** [oqtane/oqtane.framework#1621] [General]
If an Oqtane (2.1) installation has a additional tenant and there are users on both tenants with the same name (Username on the User table), logging on either of the tenants with that username gives the user access to the other tenant (even if the passwords of the two users are different).

That can't be what's intended, is it?

I presume the authorisation token needs to be verified against the Site/Tenant as well as the User table.

Or am I missing something - it's late on Friday and I'm a bit frazzled from the week...

**Comment by sbwalker** [comment]
Oqtane supports 2 types of tenancy - shared and isolated. Shared tenancy is the ability to have multiple sites in the same tenant ( ie. same database ). Isolated tenancy is the ability to physically separate sites in different tenants ( ie. different databases ). There is only a single user table per tenant. So if you are using shared tenancy, a user can have access to multiple sites within the same tenant.

Can you please clarify if the behavior you are describing is happening in the shared or isolated tenancy model?

**Reply by rossmcphee** [reply]
My tenant is isolated, I have two databases.

Here are the relevant tables:
![image](https://user-images.githubusercontent.com/63234582/130415966-a3a14378-6419-4a78-bf9c-f89353d11873.png)

I tested this with a new clone of Oqtane 2.1 Release. I've not tried it yet on 2.2.

**Reply by sbwalker** [reply]
Both sets of queries were executed against the same database in the screen shot above... I think you intended to execute them on 2 different databases? It would helpful to see the data for both databases in order to explain the behavior.

**Reply by rossmcphee** [reply]
Sorry, I had originally planned to show you both databases. Here's the correct screenshot:
![image](https://user-images.githubusercontent.com/63234582/130444773-39bf10ce-7d03-406f-a45b-7b714688f565.png)

**Comment by sbwalker** [comment]
Ok, so you have 2 tenants ( 2 databases ) - and 1 site in each database. As I mentioned above, each tenant has its own User table. And I can see from you screen shot that you have created the same 2 users in both of your tenants - host and rosshost. Note that the credentials for these users ( ie. password, etc... ) are stored independently in each tenant - so if you change them in one tenant they will NOT be replicated to the other. So this means that although the user accounts exist in both tenants, they can actually have a different password in tenant1 and tenant2. The other thing I notice is that you are using the same domain for both aliases - localhost:44357. Since cookies are linked to a domain ( and not to a subfolder ) it means that the same auth cookie will be shared when you are on localhost:44357 as well as localhost:44357/ross. 

So let's consider a scenario... let's say that the 'host' user has a password of 'password1" in tenant1 and a password of 'password2' in tenant2. When you browse to localhost:44357 and you login you would need to use 'host' and 'password1' as those are the valid credentials for tenant1. Then if you browse to localhost:44357/ross you will still be logged in as 'host' as you are still on the SAME domain ( the system will currently not force you to re-authenticate ). If you were using actual differentiated domains ( ie. domain1.com and domain2.com ) then the cookie would not be shared and you would be forced to authenticate on tenant2 ( using 'host' and 'password2' ).

Does this describe your scenario? If so, then the behavior is related to the usage of the same domain for multiple isolated tenants. It may be possible for the system to recognize that that the url changed from localhost:44357 to localhost:44357/ross and force a logout - however that would not be applicable in the case where someone is using shared tenancy.

**Reply by rossmcphee** [reply]
Thanks - that's the issue!

We were considering having separate domains for all of our customers but were concerned about he overhead involved in the routing etc. set up.  I hadn't realised that would not play well with the authentication token.
